pkg/render/render.go

```diff
@@ -56,7 +57,7 @@ func ToSVG(graph *graph.Client, cache *cache.Cache, repo string, cluster bool) (
 // ToDOT takes a GoLang repo as input and returns a DOT dependency graph
 func ToDOT(graph *graph.Client, cache *cache.Cache, repo string, cluster bool) (string, error) {
 	dot, err := cache.GetDOT(repo, cluster)
-	if err == nil {
+	if err == nil && !strings.HasPrefix(dot, "Failed") {
```

Comment:
nice catch! was there a particular repo that you were running into this issue with?

Reply:
Locally, with all repo's, not even default one worked. Live, I tried `github.com/go-micro/go-micro`

Reply:
> Locally, with all repo's, not even default one worked.

Yikes! Were `graphviz`/`dot` and `goda` installed? Let me know if I can help.

I found a good example of a repo that always fails:
`github.com/rocketlaunchr/dataframe-go`

Digging a bit further, I wanted to understand why the cache was being poisoned with messages like `Failed to render dot: ...`. On the live site I found other invalid cache values starting with `<!DOCTYPE ... `. I realized I wasn't properly handling non-200 responses from `/graph` POST requests. So, rather than trying to workaround a poisoned cache, I think we should try to prevent the poisoning in the first place. Let me know what you think about this approach:

```diff
diff --git a/pkg/graph/client.go b/pkg/graph/client.go
index bf96912..f1f7735 100644
--- a/pkg/graph/client.go
+++ b/pkg/graph/client.go
@@ -3,6 +3,7 @@ package graph
 import (
        "bytes"
        "encoding/json"
+       "errors"
        "fmt"
        "io/ioutil"
        "net/http"
@@ -85,7 +86,14 @@ func (c *Client) Get(repo string, cluster bool) (string, error) {
                return "", err
        }

-       c.log.Debugf("POST Response: %s (%d bytes)", string(body), len(respBody))
+       debugStr := fmt.Sprintf("POST response[%d] (%d bytes): %s ", resp.StatusCode, len(respBody), string(body))
+       c.log.Debug(debugStr)
+
+       if resp.StatusCode != http.StatusOK {
+               err := errors.New(debugStr)
+               httpErrors.WithLabelValues(err.Error()).Inc()
+               return "", err
+       }

        return string(respBody), nil
 }
diff --git a/pkg/graph/goda.go b/pkg/graph/goda.go
index 048e086..1ff62ab 100644
--- a/pkg/graph/goda.go
+++ b/pkg/graph/goda.go
@@ -81,7 +81,7 @@ func dirToDot(dir string, cluster bool) (string, error) {
        log.Debugf("running goda: %s", cmd)
        err := cmd.Run()
        if err != nil {
-               log.Errorf("goda cmd failed: %s", err)
+               log.Errorf("goda cmd failed [%s]: %s", err, stderr.String())
                return "", err
        }
```

Testing this against `github.com/rocketlaunchr/dataframe-go`, it correctly does not poison the cache, and also logs a more helpful error message:

```log
ERRO[0325] goda cmd failed: exit status 1: go [-e -json -compiled=true -test=false -export=false -deps=true -find=false -- ./...]: exit status 1: go: cloud.google.com/go@v0.57.0: missing go.sum entry; to add it:
	go mod download cloud.google.com/go
go: cloud.google.com/go@v0.57.0: missing go.sum entry; to add it:
	go mod download cloud.google.com/go

ERRO[0325] Failed request for [/graph]: [500] Message: [Failed to render dot: github.com/rocketlaunchr/dataframe-go] Error: [exit status 1]
DEBU[0325] POST response[500] (59 bytes): {"repo":"github.com/rocketlaunchr/dataframe-go","cluster":true}   graphclient="http://localhost:8889/graph"
ERRO[0325] error generating dot: POST response[500] (59 bytes): {"repo":"github.com/rocketlaunchr/dataframe-go","cluster":true}
ERRO[0325] Failed request for [/graph]: [500] Message: [Failed to render github.com/rocketlaunchr/dataframe-go to .svg] Error: [POST response[500] (59 bytes): {"repo":"github.com/rocketlaunchr/dataframe-go","cluster":true} ]
```

also, for now, i've cleaned out all the bad values in the cache on the live site. i'll give it another pass once we get this PR sorted and deployed.

Reply:
https://github.com/siggy/gographs/commit/8acb8085488fffff4869d2a44896e29d14dc0dcf merged and deployed to prod. thanks for motivating all this @Davincible !

Reply:
> Arguably you could simplify a bit by replacing errors.New with fmt.Errorf and put in the debug string directly.

ah, I did consider this:

```go
debugErr := fmt.Errorf("POST response[%d] (%d bytes): %s ", resp.StatusCode, len(respBody), string(respBody))
c.log.Debug(debugErr)

if resp.StatusCode != http.StatusOK {
	httpErrors.WithLabelValues(debugErr.Error()).Inc()
	return "", debugErr
}
```

my reluctance with this was `debugErr` is getting created regardless of whether an error is actually happening, and i found that a bit misleading.